More refactor of the topo package.

alainjobart opened the pull request:
A few changes:

* Moving the Lock function down to Backend, so it can be shared between keyspace and shard. Refactoring the Lock method in each implementation.
* GetKnownCells is now implemented by topo.Server.

Next steps are to move the cell map into topo.Server.

pivanof commented on go/vt/vtctld/explorer_test.go:
```diff
@@ -212,7 +209,8 @@ func TestHandleBadPath(t *testing.T) {
 
 	ts := memorytopo.New(cells...)
-	ex := newBackendExplorer(ts)
+	tts := topo.Server{Impl: ts}
```
Why not memorytopo.NewServer()?

alainjobart replied:
Done.

pivanof commented on go/vt/vttablet/tabletmanager/init_tablet.go:
```diff
@@ -155,7 +155,7 @@ func (agent *ActionAgent) InitTablet(port, gRPCPort int32) error {
 		PortMap:        make(map[string]int32),
 		Keyspace:       *initKeyspace,
-		Shard:          *initShard,
+		Shard:          shard,
```
How is this related?

alainjobart replied:
Small topo fix from the last set of commits.

pivanof replied:
I don't understand. You are changing what's stored in the topology, which means there should be a reason for that (which is not explained here) and a test verifying that you store the correct thing, right? I don't understand to which set of commits this is related, and together with the test I think it should be extracted into a separate PR.

alainjobart replied:
I pulled this change out into:
https://github.com/youtube/vitess/pull/3417
Will merge and resolve.

pivanof commented on go/vt/vtgate/vtgateservice/vtgateservice_testing/mock_vtgateservice.go:
```diff
@@ -1,25 +1,12 @@
-*/
-
-// Automatically generated by MockGen. DO NOT EDIT!
+// Code generated by MockGen. DO NOT EDIT.
```
All these mock regenerations are not related to locking, right? Why are they here?

alainjobart replied:
Because I needed to regenerate the mocks, and they all come as one. I used a different commit for the others, so it's clearer.

pivanof replied:
So you can regenerate the mocks with all the changes unrelated to this PR as a separate PR, merge them, and then this PR will contain only changes in the mocks that are related to it, right?

pivanof commented on go/vt/topo/consultopo/lock.go:
```diff
@@ -20,20 +20,41 @@ import (
+	l, err := c.client.LockOpts(&api.LockOptions{
 		Key:   lockPath,
-		Value: []byte(contents),
+		Value: []byte("lock"),
```
Why did you abandon the idea of writing contents into the lock?

alainjobart replied:
I re-added it. I thought it would be simpler, but the contents is useful in a lot of topo implementations.

alainjobart commented on go/vt/topo/backend.go:
```diff
@@ -167,9 +179,10 @@ type Version interface {
 type LockDescriptor interface {
-	// String returns a text representation of the lock.
-	String() string
+	// Unlock frees the lock.
```
Done.

alainjobart commented on go/vt/topo/consultopo/lock.go:
```diff
@@ -20,20 +20,41 @@ import (
+		if err == topo.ErrNoNode {
+			return nil, err
+		}
+		return nil, fmt.Errorf("cannot ListDir(%v,%v) before locking", cell, dirPath)
```
Done.

alainjobart commented on go/vt/topo/consultopo/lock.go:
```diff
@@ -73,21 +94,24 @@ func (s *Server) lock(ctx context.Context, c *cellClient, lockPath, contents str
+	return ld.s.unlock(ctx, ld.c, ld.lockPath)
+}
 
+// unlock releases a lock acquired by lock() on the given directory.
```
Done.

alainjobart commented on go/vt/topo/etcd2topo/lock.go:
```diff
@@ -109,101 +108,96 @@ func (s *Server) waitOnLastRev(ctx context.Context, nodePath string, revision in
 
-func (s *Server) lock(ctx context.Context, nodePath string, contents string) (string, error) {
+// ectdLockDescriptor implements topo.LockDescriptor.
+type ectdLockDescriptor struct {
```
Done.

alainjobart commented on go/vt/topo/etcd2topo/lock.go:
```diff
@@ -109,101 +108,96 @@ func (s *Server) waitOnLastRev(ctx context.Context, nodePath string, revision in
+	return s.lock(ctx, cell, dirPath, "lock")
+}
+
+// lock is used by both Lock()  and master election.
```
Done.

alainjobart commented on go/vt/topo/helpers/tee.go:
```diff
@@ -154,114 +143,56 @@ func (tee *Tee) Watch(ctx context.Context, cell, filePath string) (*topo.WatchDa
-		if err := tee.lockFirst.UnlockShardForAction(ctx, keyspace, shard, pLockPath, "{}"); err != nil {
-			log.Warningf("Failed to unlock lockFirst shard after failed lockSecond lock for %v/%v", keyspace, shard)
+		if err := fLD.Unlock(ctx); err != nil {
+			log.Warningf("Failed to unlock lockFirst after failed lockSecond lock for %v %v", cell, dirPath)
```
Done.

alainjobart commented on go/vt/topo/helpers/tee.go:
```diff
@@ -154,114 +143,56 @@ func (tee *Tee) Watch(ctx context.Context, cell, filePath string) (*topo.WatchDa
-		if perr != nil {
-			log.Warningf("Secondary UnlockShardForAction(%v/%v, %v) failed: %v", keyspace, shard, sLockPath, serr)
+		if ferr != nil {
+			log.Warningf("First Unlock(%v) failed: %v", ld.dirPath, ferr)
```
Done.

alainjobart commented on go/vt/topo/memorytopo/lock.go:
```diff
@@ -56,27 +67,33 @@ func (mt *MemoryTopo) lock(ctx context.Context, nodePath string, contents string
+	return ld.mt.unlock(ctx, ld.cell, ld.nodePath)
+}
 
+// unlock is used by Unlock().
```
Done.

alainjobart commented on go/vt/topo/test/lock.go:
```diff
@@ -69,50 +70,52 @@ func checkKeyspaceLockTimeout(ctx context.Context, t *testing.T, ts topo.Impl) {
-	if _, err := ts.LockKeyspaceForAction(ctx, "test_keyspace_666", "fake-content"); err == nil {
-		t.Fatalf("LockKeyspaceForAction(test_keyspace_666) worked for non-existing keyspace")
+// checkLockMissing makes sure we can't lock a non-existing directory.
+func checkLockMissing(ctx context.Context, t *testing.T, ts topo.Impl) {
```
Because the Lock() interface is on the Impl/Backend, not the ts, that has LockKeyspace.

pivanof replied:
But checlLockTimeout() above also verifies behavior of Lock() function and it accepts topo.Server. They verify the same Lock() function, right?

alainjobart replied:
Ah I see, fixed it with changing the signature of the other method. This will be changed again when I move the CellMap, most likely.

alainjobart commented on go/vt/topo/zk2topo/lock.go:
```diff
@@ -29,25 +29,31 @@ import (
-func (zs *Server) lockForAction(ctx context.Context, locksDir, contents string) (string, error) {
-	conn, root, err := zs.connForCell(ctx, topo.GlobalCell)
+// zsLockDescriptor implements topo.LockDescriptor.
+type zsLockDescriptor struct {
```
Done.

alainjobart commented on go/vt/topo/zk2topo/lock.go:
```diff
@@ -29,25 +29,31 @@ import (
 
 	// Create the locks path, possibly creating the parent.
-	locksPath, err := CreateRecursive(ctx, conn, locksDir, []byte(contents), zk.FlagSequence|zk.FlagEphemeral, zk.WorldACL(PermFile), 1)
+	lockPath, err := CreateRecursive(ctx, conn, locksDir, []byte("lock"), zk.FlagSequence|zk.FlagEphemeral, zk.WorldACL(PermFile), 1)
```
Done.

alainjobart commented on go/vt/topo/locks.go:
```diff
@@ -45,7 +45,6 @@ var (
 )
 
 // Lock describes a long-running lock on a keyspace or a shard.
-// It needs to be public as we JSON-serialize it.
```
Using it to generate the contents again, so re-added the comment.

pivanof commented on go/vt/topo/helpers/tee.go:
```diff
@@ -188,7 +190,7 @@ func (ld *teeTopoLockDescriptor) Unlock(ctx context.Context) error {
 	if serr != nil {
 		if ferr != nil {
-			log.Warningf("First Unlock(%v) failed: %v", ld.dirPath, ferr)
+			log.Warningf("First Unlock(%v, %v) failed: %v", ld.cell, ld.dirPath, ferr)
```
Note that you've got inconsistency in error messages: in some places cell and path are separated just by space, in some they are separated by comma. I'd personally prefer the latter format (with the comma), but in any case it should be consistent across all the messages.

alainjobart commented on go/vt/topo/etcd2topo/lock.go:
```diff
@@ -109,101 +108,96 @@ func (s *Server) waitOnLastRev(ctx context.Context, nodePath string, revision in
+// ectdLockDescriptor implements topo.LockDescriptor.
+type ectdLockDescriptor struct {
+	s       *Server
+	cell    string
```
I also fixed this, forgot to comment.

pivanof commented on go/vt/topo/etcd2topo/lock.go:
```diff
@@ -109,101 +108,81 @@ func (s *Server) waitOnLastRev(ctx context.Context, nodePath string, revision in
+	return s.lock(ctx, cell, dirPath, contents)
+}
+
+// lock is used by both Lock() and master election.
```
Mind removing this non-useful comment as well?

alainjobart replied:
This one is useful, it justifies why this is just not directly inside the Lock method.

sougou commented:
LGTM
I assume pavel's comments are addressed?
You also need to resolve the merge conflict

alainjobart commented:
OK this is ready for final review now. Re-based to master, fixed conflicts. I think I addressed all comments (which is somewhat hard to be sure of, since this review tool is not that good).

@pivanof PTAL.

pivanof commented on go/vt/topo/helpers/tee.go:
```diff
@@ -154,114 +143,58 @@ func (tee *Tee) Watch(ctx context.Context, cell, filePath string) (*topo.WatchDa
-		if err := tee.lockFirst.UnlockShardForAction(ctx, keyspace, shard, pLockPath, "{}"); err != nil {
-			log.Warningf("Failed to unlock lockFirst shard after failed lockSecond lock for %v/%v", keyspace, shard)
+		if err := fLD.Unlock(ctx); err != nil {
+			log.Warningf("Failed to unlock lockFirst after failed lockSecond lock for %v %v: %v", cell, dirPath, err)
```
Still no comma here.
